packages/tools/client-debugger/client-debugger/src/test/FluidDevtools.test.ts

```diff
@@ -0,0 +1,86 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { expect } from "chai";
+
+import { FluidDevtools } from "../FluidDevtools";
+import { ContainerDevtoolsProps } from "../ContainerDevtools";
+import { createMockContainer } from "./Utilities";
+
+// TODOs:
+// - Test window messaging
+
+describe("ClientDebugger unit tests", () => {
+	it("Container change events", () => {
+		const devtools = new FluidDevtools();
+
+		let containerRegistered = false;
+		let containerDevtoolsClosed = false;
+
+		devtools.on("containerRegistered", () => {
+			containerRegistered = true;
+		});
+		devtools.on("containerDevtoolsClosed", () => {
+			containerDevtoolsClosed = true;
+		});
+
+		expect(devtools.getAllContainerDevtools().length).to.equal(0);
+
+		const container = createMockContainer();
+		const containerId = "test-container-id";
+		const containerProps: ContainerDevtoolsProps = {
+			containerId,
+			container,
+		};
+		devtools.registerContainer(containerProps);
+
+		// eslint-disable-next-line @typescript-eslint/no-unused-expressions
+		expect(containerRegistered).to.be.true;
+		expect(devtools.getAllContainerDevtools().length).to.equal(1);
+
+		const containerDevtools = devtools.getContainerDevtools(containerId);
+		// eslint-disable-next-line @typescript-eslint/no-unused-expressions
+		expect(containerDevtools).to.not.be.undefined;
+		// eslint-disable-next-line @typescript-eslint/no-unused-expressions, @typescript-eslint/no-non-null-assertion
+		expect(containerDevtools!.disposed).to.be.false;
+
+		devtools.closeContainerDevtools(containerId);
+
+		// eslint-disable-next-line @typescript-eslint/no-unused-expressions
+		expect(containerDevtoolsClosed).to.be.true;
+		expect(devtools.getAllContainerDevtools().length).to.equal(0);
+		// eslint-disable-next-line @typescript-eslint/no-unused-expressions, @typescript-eslint/no-non-null-assertion
+		expect(containerDevtools!.disposed).to.be.true;
+
+		devtools.dispose();
+	});
+
+	it("Disposal", () => {
+		const devtools = new FluidDevtools();
+
+		let devtoolsClosed = false;
+
+		devtools.on("devtoolsDisposed", () => {
+			devtoolsClosed = true;
+		});
+
+		devtools.dispose();
+
+		// eslint-disable-next-line @typescript-eslint/no-unused-expressions
+		expect(devtoolsClosed).to.be.true;
+		// eslint-disable-next-line @typescript-eslint/no-unused-expressions
+		expect(devtools.disposed).to.be.true;
+
+		const container = createMockContainer();
+		const containerId = "test-container-id";
+		const containerProps: ContainerDevtoolsProps = {
+			containerId,
+			container,
+		};
+
+		expect(() => devtools.registerContainer(containerProps)).to.throw();
+		expect(() => devtools.closeContainerDevtools(containerId)).to.throw();
```

Comment:
That's fair. I'm always torn on this one personally, since in some ways it would be nice to not care if the specifics of an error change under the hood (for cases where we aren't trying to handle them). But I'm also happy to err on the side of being explicit.

Reply:
I think of thrown errors that we don't handle as part of the public API surface, since consumers might actually be doing that handling for specific errors; it's brittle (especially if we don't have error codes or similar so consumers can only target a particular error with a substring match on the message), but it could help us catch an unintended behavior change from "this thing is already disposed" to "internal state is invalid in some other weird way", for example. Just my 2 cents.